math/mathcore/inc/Math/CladDerivator.h

```diff
@@ -19,420 +19,215 @@
-      result[0] += 1;
-    else
-      result[1] += 1;
-  }
-
-  template <typename T>
-  void Min_darg0_grad(T a, T b, T* result) {
-    return;
-  }
-
-  template <typename T>
-  void Min_darg1_grad(T a, T b, T* result) {
-    return;
-  }
-
-  template <typename T>
-  T Power_darg0(T x, T y) {
-    return y * TMath::Power(x, y - 1);
-  }
-
-  template <typename T>
-  T Power_darg0_darg0(T x, T y) {
-    return y * (y - 1) * TMath::Power(x, y - 2);
-  }
-
-  template <typename T>
-  Double_t Power_darg0_darg1(T x, T y) {
-    return TMath::Power(x, y - 1) * (y * TMath::Log(x) + 1);
-  }
-
-  template <typename T>
-  Double_t Power_darg1(T x, T y) {
-    return TMath::Power(x, y) * TMath::Log(x);
-  }
-
-  template <typename T>
-  Double_t Power_darg1_darg0(T x, T y) {
-    return TMath::Power(x, y - 1) * (y * TMath::Log(x) + 1);
-  }
-
-  template <typename T>
-  Double_t Power_darg1_darg1(T x, T y) {
-    return TMath::Power(x, y) * TMath::Sq(TMath::Log(x));
-  }
-
-  template <typename T>
-  Double_t Power_grad(T x, T y, Double_t* result) {
-    T t = TMath::Power(x, y - 1);
-    result[0] += y * t;
-    result[1] += x * t * TMath::Log(x);
-  }
-
-  template <typename T>
-  Double_t Power_darg0_grad(T x, T y, Double_t* result) {
-    T t = TMath::Power(x, y - 2);
-    result[0] += y * (y - 1) * t;
-    result[1] += x * t * (y * TMath::Log(x) + 1);
-  }
-
-  template <typename T>
-  Double_t Power_darg1_grad(T x, T y, Double_t* result) {
-    T t = TMath::Power(x, y - 1);
-    result[0] += t * (y * TMath::Log(x) + 1);
-    result[1] += x * t * TMath::Sq(TMath::Log(x));
-  }
-
-  template <typename T>
-  Double_t Sin_darg0(T d) {
-    return TMath::Cos(d);
-  }
-
-  template <typename T>
-  Double_t Sin_darg0_darg0(T d) {
-    return -TMath::Sin(d);
-  }
-
-  template <typename T>
-  Double_t SinH_darg0(T d) {
-    return TMath::CosH(d);
-  }
-
-  template <typename T>
-  Double_t SinH_darg0_darg0(T d) {
-    return TMath::SinH(d);
-  }
-
-  template <typename T>
-  T Sq_darg0(T d) {
-    return 2 * d;
-  }
-
-  template <typename T>
-  T Sq_darg0_darg0(T d) {
-    return 2;
-  }
-
-  template <typename T>
-  Double_t Sqrt_darg0(T d) {
-    return 0.5 / TMath::Sqrt(d);
-  }
-
-  template <typename T>
-  Double_t Sqrt_darg0_darg0(T d) {
-    return 0.25 / (TMath::Sqrt(d) * d);
-  }
-
-  template <typename T>
-  Double_t Tan_darg0(T d) {
-    return 1./ TMath::Sq(TMath::Cos(d));
-  }
-
-  template <typename T>
-  Double_t Tan_darg0_darg0(T d) {
-    return 2. * TMath::Sin(d) / (TMath::Cos(d) * TMath::Cos(d) * TMath::Cos(d));
-  }
-
-  template <typename T>
-  Double_t TanH_darg0(T d) {
-    return 1./ TMath::Sq(TMath::CosH(d));
-  }
-
-  template <typename T>
-  Double_t TanH_darg0_darg0(T d) {
-    return -2. * TMath::SinH(d) / (TMath::CosH(d) * TMath::CosH(d) * TMath::CosH(d));
-  }
+namespace TMath {
+template <typename T>
+ValueAndPushforward<T, T> Abs_pushforward(T x, T d_x)
+{
+   return {::TMath::Abs(x), ((x < 0) ? -1 : 1) * d_x};
```

Comment:
I would not use `TMath::Abs` but `std::abs`. 
Similar for all Math functions

Reply:
I don't think we should use std math functions here. Sure, `std` math functions are more robust and scalable. If a user decides to use `TMath` functions then Clad should use `TMath` functions. It is not the place of Clad to create any dispute against the user code.

The pushforward function strategy returns two values for a function `fn`. Its value at the given argument values and its *pushforward* with respect to the specified argument at the given argument values. If we use `std` function in the pushforward custom derivative code, then the resulting derivative code would effectively be like the user had used `std` functions instead of `TMath` function. Clad shouldn't go against the wishes of the user.

Clad should indeed provide custom derivatives of both `std` and `TMath`. So users can use whichever of the two based on their requirements. Clad also provides inbuilt support of custom derivatives of `std` functions, so they can be used as well.